lib/bazel/commands/commands.go

```diff
@@ -87,6 +91,84 @@ func NewAffectedTargetsList(parent *AffectedTargets) *AffectedTargetsList {
+	gitRoot, err := findGitRoot()
+	if err != nil {
+		return fmt.Errorf("can't find git repo root: %v", err)
+	}
+
+	// Create temporary worktrees in which to execute bazel commands.
+	// If the end commit is not provided, use the current git directory as the end
+	// worktree, which will include uncommitted local changes.
+	startTree, err := git.NewTempWorktree(gitRoot, c.parent.Start)
+	if err != nil {
+		return fmt.Errorf("can't generate worktree for committish %q: %v", c.parent.Start, err)
+	}
+	defer startTree.Close()
+
+	endTreePath := gitRoot
+	if c.parent.End != "" {
+		endTree, err := git.NewTempWorktree(gitRoot, c.parent.End)
+		if err != nil {
+			return fmt.Errorf("can't generate worktree for committish %q: %v", c.parent.End, err)
+		}
+		defer endTree.Close()
+		endTreePath = endTree.Root()
+	}
+
+	// Open the bazel workspaces, using a well-known output_base. Since the
+	// temporary worktrees created above will have a different path on every
+	// invocation, by default bazel will create a new cache directory for them,
+	// re-download all dependencies, etc. which is both slow and will eventually
+	// fill up the disk. Reusing an output base location will ensure that at most
+	// only two are created for the purposes of this subcommand.
+	startOutputBase, err := cacheDir("affected_targets/start")
+	if err != nil {
+		return fmt.Errorf("failed to create output_base: %v", err)
+	}
+	endOutputBase, err := cacheDir("affected_targets/end")
+	if err != nil {
+		return fmt.Errorf("failed to create output_base: %v", err)
+	}
+	startWorkspace, err := bazel.OpenWorkspace(startTree.Root(), bazel.WithOutputBase(startOutputBase))
+	if err != nil {
+		return fmt.Errorf("failed to open bazel workspace for committish %q: %v", c.parent.Start, err)
+	}
+	endWorkspace, err := bazel.OpenWorkspace(endTreePath, bazel.WithOutputBase(endOutputBase))
+	if err != nil {
+		return fmt.Errorf("failed to open bazel workspace: %v", err)
+	}
+
+	// Get all target info for both VCS time points.
+	targets, err := startWorkspace.Query("deps(//...)", bazel.WithKeepGoing(), bazel.WithUnorderedOutput())
+	if err != nil {
+		return fmt.Errorf("failed to query deps for start point: %v", err)
+	}
+	// TODO(scott): Replace with logging
+	fmt.Fprintf(os.Stderr, "Processed %d targets at start point\n", len(targets))
+
+	targets, err = endWorkspace.Query("deps(//...)", bazel.WithKeepGoing(), bazel.WithUnorderedOutput())
+	if err != nil {
+		return fmt.Errorf("failed to query deps for end point: %v", err)
+	}
+	// TODO(scott): Replace with logging
+	fmt.Fprintf(os.Stderr, "Processed %d targets at end point\n", len(targets))
+
 	return fmt.Errorf("not yet implemented")
 }
 
+func cacheDir(subDir string) (string, error) {
+	d, err := os.UserCacheDir()
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(d, "enkit/bazel", subDir), nil
+}
+
+func findGitRoot() (string, error) {
+	bazelWorkspace := os.Getenv("BUILD_WORKSPACE_DIRECTORY")
```

Comment:
Note that the BUILD_WORKSPACE_DIRECTORY may not match the git root. WORKSPACE may start in a subdirectory, or even have multiple WORKSPACEs in the same repo.

Let's not worry about handling this complexity in code, but I'd rename the function to match the intent.

Intent is to find workspace root? call it findWorkspaceRoot, use code like I have at the bottom of this code.

Intent is to find the git root? remove the check for BUILD_WORKSPACE_DIRECTORY,  invoke the comment in my next comment.

Here's a snippet of code I wrote in the past to find the WORKSPACE root:
```
import (
        "github.com/bazelbuild/buildtools/wspace"
)

func WorkspaceRoot() string {
        root := os.Getenv("BUILD_WORKSPACE_DIRECTORY")
        if root != "" {
                return root
        }

        root, _ = wspace.FindWorkspaceRoot("")
        return root
}
```

Reply:
Hmm the intent is complected.

There is an assumption that the git root and the workspace root are one and the same for repos in which this will be run (enkit and internal, do you foresee a scenario where this assumption breaks down?)

I guess the "correct" method is both: we need to find the git root to execute the worktree commands, and then find the workspace root of the resulting worktrees.

`BUILD_WORKSPACE_DIRECTORY` is still needed if we expect people to invoke `enkit` via `bazel run` (do we?) because otherwise PWD will be completely outside the git tree.